victor/src/layout/mod.rs

```diff
@@ -19,43 +20,82 @@ impl crate::dom::Document {
         layout_document(&box_tree, viewport)
     }
 }
+
 fn layout_document(
     box_tree: &BoxTreeRoot,
     viewport: crate::primitives::Size<crate::primitives::CssPx>,
 ) -> Vec<Fragment> {
+    let inline_size = Length { px: viewport.width };
+
     // FIXME: use the document’s mode:
     // https://drafts.csswg.org/css-writing-modes/#principal-flow
     let initial_containing_block = ContainingBlock {
-        inline_size: Length { px: viewport.width },
+        inline_start_from_absolute_containing_block: Length::zero(),
+        inline_size,
         block_size: Some(Length {
             px: viewport.height,
         }),
         mode: (WritingMode::HorizontalTb, Direction::Ltr),
     };
 
-    let (fragments, _) = box_tree.layout(&initial_containing_block);
+    let (mut fragments, absolutely_positioned_fragments, _) =
+        box_tree.layout(&initial_containing_block, &initial_containing_block, 0);
+    fragments.extend(
+        absolutely_positioned_fragments
+            .into_iter()
+            .map(|fragment| Fragment::Box(fragment.contents)),
+    );
     fragments
 }
 
+struct AbsoluteContext<'a> {
+    containing_block: &'a ContainingBlock,
+    absolutely_positioned_fragments: Mutex<Vec<AbsolutelyPositionedFragment>>,
+}
+
+struct AbsolutelyPositionedFragment {
+    index: usize,
+    uses_static_block_position: bool,
+    contents: BoxFragment,
+}
+
 struct ContainingBlock {
+    inline_start_from_absolute_containing_block: Length,
     inline_size: Length,
     block_size: Option<Length>,
     mode: (WritingMode, Direction),
 }
 
 impl BlockFormattingContext {
-    fn layout(&self, containing_block: &ContainingBlock) -> (Vec<Fragment>, Length) {
-        self.0.layout(containing_block)
+    fn layout(
+        &self,
+        absolute_containing_block: &ContainingBlock,
+        containing_block: &ContainingBlock,
+        index: usize,
+    ) -> (Vec<Fragment>, Vec<AbsolutelyPositionedFragment>, Length) {
+        self.0
+            .layout(absolute_containing_block, containing_block, index)
     }
 }
 
 impl BlockContainer {
-    fn layout(&self, containing_block: &ContainingBlock) -> (Vec<Fragment>, Length) {
+    fn layout(
+        &self,
+        absolute_containing_block: &ContainingBlock,
+        containing_block: &ContainingBlock,
+        index: usize,
+    ) -> (Vec<Fragment>, Vec<AbsolutelyPositionedFragment>, Length) {
         match self {
             BlockContainer::BlockLevelBoxes(child_boxes) => {
+                let mut absolute_context = AbsoluteContext {
```

Comment:
Given that one `AbsoluteContext` is created for each nested block container (rather than one per “containing block for abspos descendants” as I was thinking when we first discussed this) I wonder if the `Mutex` and sorting are necessary at all, or if we could have the same data be part of the return type of `BlockLevelBox::layout`.

Reply:
I did it that way because then the `Vec<Fragment>` returned by `BlockContainer::layout` doesn't need to be transformed into something else (to not get a `Vec<(Fragment, Vec<AbsposFragment>)>` or something like that), under the assumption that there are few abspos fragments anyway.

As for "one containing block for abspos descendants", I could reduce the number of times we need to go through a temporary mutex and vector if I knew whether any given element has multiple children with absolutely-positioned elements, but I can't have only one per containing block for abspos descendants because `BlockContainer::layout` does its job in parallel so the pushes to `absolutely_positioned_fragments` happen in parallel, so each element is responsible for keeping their absolutely-positioned descendants ordered.